src/include/daos_srv/ad_mem.h

```diff
@@ -0,0 +1,69 @@
+#ifndef __DAOS_AD_HOC_MEM_H__
+#define __DAOS_AD_HOC_MEM_H__
+
+#include <daos/mem.h>
+
+/* the memory region (blob) manged by ad-hoc allocator */
+struct ad_blob;
+
+/* Ad-hoc mem transaction handle */
+struct ad_tx {
+	struct ad_blob		*tx_blob;
+	uint64_t		 tx_id;
+	bool			 tx_redo_only;
+	d_list_t		 tx_redo;
+	d_list_t		 tx_undo;
+};
+
+/** start a ad-hoc memory transaction */
+int
+ad_tx_begin(struct ad_blob *blob, struct ad_tx *tx);
+
+/** complete a ad-hoc memory transaction */
+int
+ad_tx_end(struct ad_tx *tx, int err);
+
+/**
+ * declare this storage region is going to be changed (for redo and undo), it does memset
+ * the region if @reset is true.
+ */
+int
+ad_tx_touch_region(void *addr, daos_size_t size, bool reset, struct ad_tx *tx);
+
+/** finished the change */
+int
+ad_tx_touch_done(struct ad_tx *tx);
+
+/**'
+ * copy data from buffer @ptr to storage address @addr, both old and new data will be saved
+ * for TX redo and undo.
+ */
+int
+ad_tx_copy(struct ad_tx *tx, void *addr, daos_size_t size, void *ptr, bool ptr_only);
```

Comment:
probably don't need the "tx" parameter in these set of APIs, similar as PMDK _tx APIs don't with tx argument, and UMEM tx API also don't take a tx argument.
probably can use a thread local variable for the tx (alloc it in ad_tx_begin, close/commit it in ad_tx_end) and get/query it internally inside ad_tx APIs.
And we probably can with external TX (explicit start by ad_tx_begin), and internal TX (for atomic allocate and publish not in external TX).

Reply:
seems also fine to with tx argument, so query outside the API if needed. sorry, nvm.

Reply:
what mean for ptr_only?

Reply:
only store the buffer address in ad_alloc() w/o copy data, bio can copy the content to WAL instead of coping twice